### PR TITLE
Merge feature-vecsim to master

### PR DESCRIPTION
**Pull request by rafie**

### CODE REVIEW

File: src/spec.c [hunk 1]
```diff
@@ -975,6 +1099,11 @@ RedisModuleString *IndexSpec_GetFormattedKey(IndexSpec *sp, const FieldSpec *fs,
       case INDEXFLD_T_TAG:
         ret = TagIndex_FormatName(&sctx, fs->name);
         break;
+      case INDEXFLD_T_VECTOR:
+        // TODO: remove the whole thing
+        // NOT NECESSARY ANYMORE - used when field were in keyspace
```
Comment (ashtul) [inline src/spec.c:1104]: clean

File: src/spec.c [hunk 2]
```diff
@@ -1196,6 +1325,7 @@ static const FieldType fieldTypeMap[] = {[IDXFLD_LEGACY_FULLTEXT] = INDEXFLD_T_F
                                          [IDXFLD_LEGACY_NUMERIC] = INDEXFLD_T_NUMERIC,
                                          [IDXFLD_LEGACY_GEO] = INDEXFLD_T_GEO,
                                          [IDXFLD_LEGACY_TAG] = INDEXFLD_T_TAG};
+                                         // CHECKED: Not related to new data types - legacy code
```
Comment (ashtul) [inline src/spec.c:1328]: remove

File: src/vector_index.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  unsigned char *vector = vf->vector;
+  switch (vf->type) {
+    case VECTOR_SIM_TOPK:
+      if (vf->isBase64) {
+        unescape((char *)vector, &vf->vecLen);
+        vector = base64_decode(vector, vf->vecLen, &outLen);
+      }
+      
+      VecSimQueryParams qParams = {.hnswRuntimeParams.efRuntime = vf->efRuntime};
+      vf->results = VecSimIndex_TopKQuery(vecsim, vector, vf->value, &qParams, BY_ID );
```
Comment (ashtul) [inline src/vector_index.c:80]: space

File: src/vector_index.c [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+void VectorFilter_InitValues(VectorFilter *vf) {
+  vf->efRuntime = HNSW_DEFAULT_EF_RT;
+}
+
+
+VectorQueryType VectorFilter_ParseType(const char *s, size_t len) {
+  if (!strncasecmp(s, "TOPK", len)) {
+    return VECTOR_SIM_TOPK;
+  } else if (!strncasecmp(s, "RANGE", len)) {
+    return VECTOR_SIM_TOPK;
```
Comment (ashtul) [inline src/vector_index.c:103]: disable `RANGE` keyword

File: src/vector_index.c [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    return VECTOR_SIM_INVALID;
+  }
+}
+
+int VectorFilter_Validate(const VectorFilter *vf, QueryError *status) {
+    if (vf->type == VECTOR_SIM_INVALID) {
+      QERR_MKSYNTAXERR(status, "Invalid Vector similarity type");
+      return 0;
+    }
+    return 1;
```
Comment (ashtul) [inline src/vector_index.c:114]: spaces

**Review by ashtul — COMMENTED** [review]
TODO

File: tests/pytests/test_search_params.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+    res2 = conn.execute_command('FT.SEARCH', 'idx', '@name:(John ~$val1)', 'NOCONTENT', 'PARAMS', '2', 'val1', 'Doh')
+    env.assertEqual(res2, res1)
+
+    # FIXME: Avoid parameterization in verbatim string (whether a param is defined or not)
+    #  Parser seems OK
+    #  (need to review indexing, in previous versions the following search query was syntactically illegal)
+    # res1 = conn.execute_command('FT.SEARCH', 'idx', '@name:("$val1")', 'NOCONTENT')
+    # env.assertEqual(res1, [1L, 'key5'])
+    # res2 = conn.execute_command('FT.SEARCH', 'idx', '@name:("$val1")', 'NOCONTENT', 'PARAMS', '2', 'val1', 'Alice')
+    # env.assertEqual(res2, res1)
```
Comment (rafie) [inline tests/pytests/test_search_params.py:220]: what about this one?

Comment (oshadmi) [reply tests/pytests/test_search_params.py:220]: @rafie This is a pending completion, probably on indexing (in Search 2.0 a `$` here was causing a syntax error)

**Review by oshadmi — DISMISSED** [review]

**Review by ashtul — APPROVED** [review]

**Pull request merged**